File: apps/android/app/src/main/kotlin/com/nogeass/passalarm/domain/model/ProStatus.kt
package com.nogeass.passalarm.domain.model

enum class ProTier { FREE, PRO }
enum class ProPeriod { MONTHLY, YEARLY }
enum class ProSource { STORE, CROWDFUND, MANUAL }

data class ProStatus(
    val tier: ProTier = ProTier.FREE,
    val expiresAt: Long? = null,
    val period: ProPeriod? = null,
    val source: ProSource = ProSource.STORE,
) {
    val isPro: Boolean get() = tier == ProTier.PRO
    val isLifetime: Boolean get() = isPro && source == ProSource.CROWDFUND && expiresAt == null
    companion object {
        val Free = ProStatus()
    }
}

data class ProProduct(
    val id: String,
    val period: ProPeriod,
    val displayPrice: String,
    val pricePerMonth: String?,
    val trialText: String?,
)

File: apps/ios/PassAlarm/Domain/Entity/ProStatus.swift
import Foundation

enum ProTier: String, Sendable, Codable {
    case free, pro
}

enum ProPeriod: String, Sendable, Codable {
    case monthly, yearly
}

enum ProSource: String, Sendable, Codable {
    case store
    case crowdfund
    case manual
}

struct ProStatus: Sendable, Equatable {
    let tier: ProTier
    let expiresAt: Date?
    let period: ProPeriod?
    let source: ProSource

    var isPro: Bool { tier == .pro }
    var isLifetime: Bool { isPro && source == .crowdfund && expiresAt == nil }
    static let free = ProStatus(tier: .free, expiresAt: nil, period: nil, source: .store)

    init(tier: ProTier, expiresAt: Date?, period: ProPeriod?, source: ProSource = .store) {
        self.tier = tier
        self.expiresAt = expiresAt
        self.period = period
        self.source = source
    }
}

struct ProProduct: Identifiable, Sendable {
    let id: String
    let period: ProPeriod
    let displayPrice: String
    let pricePerMonth: String?
    let trialText: String?
}
